packages/gatsby/src/schema/__tests__/node-model.js

```diff
@@ -651,43 +741,53 @@ describe(`NodeModel`, () => {
         expect(trackedRootNode).not.toEqual(node)
       })
     })
-
-    describe(`Tracks nodes returned by queries`, () => {
-      it(`Tracks objects when running query without filter`, async () => {
-        const result = await nodeModel.runQuery({
-          query: {},
-          type: schema.getType(`Test`),
-          firstOnly: false,
+    ;[
+      { desc: `with index cache`, cb: () => new Map() }, // Avoids sift
+      { desc: `no index cache`, cb: () => undefined }, // Requires sift
+    ].forEach(({ desc, cb: createIndexCache }) => {
+      describe(`[${desc}] Tracks nodes returned by queries`, () => {
+        it(`Tracks objects when running query without filter`, async () => {
+          const result = await nodeModel.runQuery({
+            query: {},
+            type: schema.getType(`Test`),
+            firstOnly: false,
+            typedKeyValueIndexes: nodeModel.replaceTypeKeyValueCache(
+              createIndexCache()
+            ),
+          })
+
+          expect(result.length).toEqual(2)
+          expect(
+            nodeModel.findRootNodeAncestor(result[0].inlineObject)
+          ).toEqual(result[0])
+          expect(
+            nodeModel.findRootNodeAncestor(result[1].inlineObject)
+          ).toEqual(result[1])
         })
 
-        expect(result.length).toEqual(2)
-        expect(nodeModel.findRootNodeAncestor(result[0].inlineObject)).toEqual(
-          result[0]
-        )
-        expect(nodeModel.findRootNodeAncestor(result[1].inlineObject)).toEqual(
-          result[1]
-        )
-      })
-
-      it(`Tracks objects when running query with filter`, async () => {
-        const result = await nodeModel.runQuery({
-          query: {
-            filter: {
-              inlineObject: {
-                field: {
-                  eq: `fieldOfSecondNode`,
+        it(`Tracks objects when running query with filter`, async () => {
+          const result = await nodeModel.runQuery({
+            query: {
+              filter: {
+                inlineObject: {
+                  field: {
+                    eq: `fieldOfSecondNode`,
+                  },
                 },
               },
             },
-          },
-          type: schema.getType(`Test`),
-          firstOnly: false,
+            type: schema.getType(`Test`),
+            firstOnly: false,
+            typedKeyValueIndexes: nodeModel.replaceTypeKeyValueCache(
```

Comment:
Sorry for nitting, but this implies we actually use it, even though this is a undefined returning side-effect. I sugest

```
nodeModel.replaceTypeKeyValueCache()
const result = await ...
```

Reply:
Sure :)

Yeah this is where basic typing would be handy. The type would pseudo-enforce / document the return value being undefined / "unused", but rather an artifact of return-early style.

Small tangent but this was my main beef about implicit return values in coffeescreipt and es6 arrows. But that boat went tuut tuut.